core/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreRestartTest.java

```diff
@@ -257,36 +283,45 @@ public void testRestartSameKey(boolean deleteIndex) throws Throwable {
       createCacheManagers();
    }
 
-   @Test(dataProvider = "booleans")
-   public void testRestartWithRemoveOnly(boolean deleteIndex) throws Throwable {
-      cache(0, cacheName).remove("k1");
+   public void testRestartCompactorNotComplete() throws Throwable {
+      if (fileSize > 320_000) {
+         // Don't test larger as tests take way too long
+         return;
+      }
+      WaitDelegatingNonBlockingStore store = TestingUtil.getFirstStoreWait(cache(0, cacheName));
+
+      Compactor compactor = TestingUtil.extractField(store.delegate(), "compactor");
+      cache(0, cacheName).remove("removed-key");
+      int size = 0;
+      // Insert until compactor has filled a file - which we can force compaction on below
+      while (compactor.getFileStats().isEmpty()) {
+         cache(0, cacheName).put("key-" + size, "value-" + size);
+         size++;
+      }
 
-      Supplier<Long> statsUsedSize = () -> {
-         NonBlockingSoftIndexFileStore sifs = TestingUtil.getFirstStore(cache(0, cacheName));
+      CheckPoint checkPoint = new CheckPoint();
+      checkPoint.triggerForever(Mocks.AFTER_RELEASE);
+      Mocks.blockingFieldMock(checkPoint, Index.class, compactor, Compactor.class, "index",
+            (stubber, index) -> stubber.when(index).handleRequest(any()));
 
-         Compactor compactor = TestingUtil.extractField(sifs, "compactor");
+      CompletionStage<Void> compactorStage = compactor.forceCompactionForAllNonLogFiles();
 
-         ConcurrentMap<Integer, Compactor.Stats> stats = compactor.getFileStats();
+      checkPoint.awaitStrict(Mocks.BEFORE_INVOCATION, 10, TimeUnit.SECONDS);
 
-         long size = 0;
-         for (Compactor.Stats stat : stats.values()) {
-            size += stat.getTotal();
-            size -= stat.getFree();
-         }
-         return size;
-      };
+      Future<Void> killFuture = fork(() -> killMember(0, cacheName));
 
-      // This will be -40 due to tracking the file has 40 bytes free, but total size is not set until file complete
-      eventuallyEquals(-40L, statsUsedSize);
+      checkPoint.triggerForever(Mocks.BEFORE_RELEASE);
 
-      killMember(0, cacheName);
+      compactorStage.toCompletableFuture().get(10, TimeUnit.SECONDS);
+
+      killFuture.get(10, TimeUnit.SECONDS);
 
-      if (deleteIndex) {
-         // Delete the index which should force it to rebuild
-         Util.recursiveFileRemove(Paths.get(tmpDirectory, "index"));
-      }
       createCacheManagers();
 
-      eventuallyEquals(0L, statsUsedSize);
+      assertNull(cache(0, cacheName).get("removed-key"));
+
+      for (int i = 0; i < size; i++) {
+         assertEquals("value-" + i, cache(0, cacheName).get("key-" + i));
+      }
```

Comment:
Is there a way to tell if the file was compacted after restart? Or maybe we don't need to check it?

Reply:
The test was more about ensuring there was no error during compaction and that the index was in fact okay. But I can try adding something for the size of data directory.